### ISSUE TITLE
Performance of nn.conv1d and keras.layers.Conv1D is low the first time any given input size is processed even if retracing is prevented! 

### ISSUE THREAD

**Issue by roebel** [tensorflow/tensorflow#54456]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):

yes

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):

Linux Ubuntu 20.04

- TensorFlow installed from (source or binary):

pip

- TensorFlow version (use command below):

tested with TF 2.4, 2.6, 2.8

- Python version:

3.7

- CUDA/cuDNN version:

Cuda Toolkit 11.3.1
CUDNN 8.3.1 

- GPU model and memory:

Tested with GeForce GTX 1050 Ti  and GeForce GTX 1080 Ti

**Describe the current behavior**

In both cases using eager mode and using a `tf.function` with `experimental_relax_shapes=True`
running tf.nn.conv1d is slow the first time a tensor of any given size is processed and the processing of a new tensor of the same size then becomes 4 times (on GPU 1050 TI) or 10 times (on GPU 1080 Ti) faster the second or further times
.
The observed behavior is a severe problem for running inference with audio signals because audio signals generally have very different sizes and therefore in a production environment the code will run only with 10% of maximum performance (on a GPU 1080 Ti) 
for the first few 100k examples until the model has seen sufficiently many lengths to achieve peak performance.

**Describe the expected behavior**

conv1d processing time should depend on the size of the input vector and not on the number of times the same size has been seen.

**Standalone code to reproduce the issue**

colab notebook is 
[here](https://colab.research.google.com/gist/roebel/5cc15bb93566d374e87da9d2c9e885f2/performance_issue.ipynb)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem.

Result running on colab with GPU

```
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
=== run 1 ===
len 10000 time: 10.187 speed 9.82kHz su (10, 9991, 100)
len 10001 time: 0.603 speed 165.93kHz su (10, 9992, 100)
len 10002 time: 0.613 speed 163.08kHz su (10, 9993, 100)
len 10003 time: 0.628 speed 159.26kHz su (10, 9994, 100)
len 10004 time: 0.587 speed 170.35kHz su (10, 9995, 100)
=== run 2 ===
len 10000 time: 0.051 speed 1946.11kHz su (10, 9991, 100)
len 10001 time: 0.050 speed 2000.30kHz su (10, 9992, 100)
len 10002 time: 0.066 speed 1518.87kHz su (10, 9993, 100)
len 10003 time: 0.056 speed 1788.66kHz su (10, 9994, 100)
len 10004 time: 0.059 speed 1705.02kHz su (10, 9995, 100)
```

*Notes:*
- You notice 10 times increase the second time the inner loop is run. If you run the same notebook again the first run will be fast as well, which indicates the GPU is caching something. 
- to have the effect again your need to restart the notebook
- the behavior is not CUDA imposed because an equivalent PyTorch script runs the first and second pass of the inner loop without speed difference.

**Comment by roebel** [comment]
I believe to have made some progress using tensorboard profiler. The first time the convolution is run on a given size the profiler displays running all the following kernels
 
- `maxwell_scudnn_128x128_relu_small_nn_v1`,  
- `implicit_convolve_sgemm`, 
- `cudnn::cnn::im2col4d_kernel`, 
- `explicit_convolve_sgemm`, 
- `fft1d_c2r_256`, 
- `maxwell_gcgemm_32x32_nt`. 
 
The second time any given size is processed only a single kernel is used. In the present case these are either  `maxwell_gcgemm_32x32_nt` or `maxwell_scudnn_128x128_relu_small_nn_v1`. So it appears tensorflow is trying to optimize by means of adaptively selecting the best kernel for each size. 

This is problematic notably for inference with audio where size of audio files can vary between 1 second (16000 samples) and 40 seconds (640000 samples) because as mentioned above the software will spend most of the time trying to optimize and will therefore need more than a day before it starts to perform optimally. The question here would be whether this trial stage can be prevented by means of manually preselecting a strategy.

**Comment by mohantym** [comment]
Hi @chunduriv ! Could you please look at this issue? Its replicating in [2.8](https://colab.sandbox.google.com/gist/mohantym/20beff230772268114e37ec0b2d9eeed/performance_issue.ipynb#scrollTo=M521fw4_sNSA) and throwing different error in [2.7](https://colab.sandbox.google.com/gist/mohantym/966d7b901aa5063ad79c278eceec9679/performance_issue.ipynb#scrollTo=0SB4KiqP_a67) and [nightly](https://colab.sandbox.google.com/gist/mohantym/d201db4c877782d9334802cf953a529f/performance_issue.ipynb#scrollTo=0SB4KiqP_a67). Thanks!

**Comment by mohantym** [comment]
@roebel ! Did you try the same in [distribution training ](https://www.tensorflow.org/guide/distributed_training#:~:text=Overview-,tf.,code%20with%20minimal%20code%20changes.)yet?

**Comment by roebel** [comment]
@mohantym Thanks for your reply. But I wonder why you suggest distribution training? I don't have a problem with training I have a problem with inference. Also for my use cases, I don't have access to multiple GPUs so I cannot use distribution inference - or could I? If these tests would be run in parallel on the same GPU this may help but would also require more memory.

**Comment by LerysG** [comment]
Hi, got a similar problem (running on GTX 1080), using conv1d layers makes my kernel died every time, as my memory GPU is fully saturated even with very low dim input shpaes. I tried tensorflow version 2.5, 2.6, 2.7 and 2.9. I am running on CUDA v11.6 and CudNN v8.4.1 (windows). I could not find any thing to avoid crashing. If anyone has an idea, your very much welcome!

**Comment by LerysG** [comment]
This thread save me : https://stackoverflow.com/questions/72356588/could-not-locate-zlibwapi-dll-please-make-sure-it-is-in-your-library-path

**Comment by sachinprasadhs** [comment]
@roebel , Could you try the solution from the above link, also check the solution from the link here https://github.com/tensorflow/tensorflow/issues/56387 which is similar to your issue, which was solved by exporting path.

**Comment by roebel** [comment]
@sachinprasadhs 

I am sorry but the report of @LerysG is completely unrelated to what I describe here. 

- I don't have crash
- I don't have a library missing.

So there is nothing I could reasonably check here.

Similarly, issue 56387 is not related at all.

The problem I describe is an implementation problem in the TensorFlow code.

**Comment by roebel** [comment]
I think I found the source code that is responsible for the issue: cuda supports a number of convolution kernels that in the current version of TensorFlow 2.9.0 are obtained by means of CudnnSupport::GetConvolveRunners here

https://github.com/tensorflow/tensorflow/blob/21368c687cafdf97fac3dd0eefaed710df0068a2/tensorflow/stream_executor/cuda/cuda_dnn.cc#L4557

Which is then used here in autotune functions

https://github.com/tensorflow/tensorflow/blob/21368c687cafdf97fac3dd0eefaed710df0068a2/tensorflow/core/kernels/conv_ops_gpu.cc#L365

It appears that each time a configuration consisting of data shape, filter shape, and maybe other parameters are encountered the cuda driver tests all of the kernels and retains the most efficient one. This is a very nice optimization for most cases, notably training with constant batch shapes, or inference with constant image sizes. For inference with audio signals the cuda implementation is testing most of the time all kernels versions. It hardly ever benefits, from the information which of the kernels is the most efficient one for any given configuration, as the same configuration is hardly ever encountered a second time.

All this reminds me of fftw3, which also uses such kind of autotuning depending on the FFT size, but there FFT sizes do not change so much, and more over it is possible to store results in form of wisdom.

I think for the present case it would be nice to affect the kernel selection explicitly. I wonder if one cannot define a maximum size above which the selected kernel would no longer be adapted. E.g assuming one would be able to select a maximum adaptation length via env variable TENSORFLOW_CONV_MAXADAPLEN and if the variable is set, one would then always run autotuning with min (shape[i],  TENSORFLOW_CONV_MAXADAPLEN). In that case the current problem could be solved if needed without any negative impact for te cases that have networks working on constant shapes (image processing).

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by roebel** [comment]
Hello, 

thanks for your comment, but this is probably a joke. Since the 2 years that I have opened the issue, and documented the exact location in the source code where the problem is located, I have not received any competent feedback. Now you ask me to open a new issue delivering updated information?

> We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue.

I cannot afford to buy the latest hardware to confirm the issue is still an issue.
In any case, the problem is not solved. Here the colab that I had produced above, but  running under TF2.17

https://colab.research.google.com/drive/1X_zJLfVboHfne7y2RC5Dp0P20CiJbopI#scrollTo=M521fw4_sNSA

Still, 3 times slower for every new input size.

I agree colab is not good for measurements.
At the end of this message you find the same test running with the most recent version I use on our cluster (TF2.16.2) - exactly as for TF2.9, for every new input size it is 10 times slower compared to the second time it is run. 

The problem is by design, but it appears the use case of constantly changing input sizes is not of central interest for the TF developers. Why should I open a new issue then.

Best


=============
tf version 2.16.2
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
=== run 1 ===
len 10000 time: 2.642 speed 37.84kHz su (10, 9991, 100)
len 10001 time: 0.216 speed 463.14kHz su (10, 9992, 100)
len 10002 time: 0.171 speed 585.71kHz su (10, 9993, 100)
len 10003 time: 0.186 speed 538.95kHz su (10, 9994, 100)
len 10004 time: 0.135 speed 738.84kHz su (10, 9995, 100)
=== run 2 ===
len 10000 time: 0.027 speed 3753.29kHz su (10, 9991, 100)
len 10001 time: 0.021 speed 4784.02kHz su (10, 9992, 100)
len 10002 time: 0.018 speed 5491.67kHz su (10, 9993, 100)
len 10003 time: 0.018 speed 5505.34kHz su (10, 9994, 100)
len 10004 time: 0.018 speed 5552.22kHz su (10, 9995, 100)
